0100-private-projects.md

```diff
@@ -0,0 +1,36 @@
+- Feature Name: Private Projects
+- Created: 2020-02-10
+- Associated Bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1602490
+
+# Summary
+
+Add an option to make a project "private", restricting who can see and interact with it.
+
+# Motivation
+
+It can be useful in some cases to be able to hide a project from users. For example, when a superuser creates a new project, they might want to make sure that everything is correctly set up before exposing it — running sync, verifying that it is enabled for the right locales, etc. In such a case, the project being not visible from non-superusers is enough. Once the superuser is done with configuration, they simply make the project public.
+
+For a different use-case, let's consider an agency working with several clients. They have a number of projects created on Pontoon, and they want to have specific localizers working on specific projects. It would be beneficial for them to able to restrict seeing a project to only a defined list of users. This requires being able to mark a project as private, and also having a configuration option to select a list of users who will be able to see, and interact, with that project. That would allow a clear separation of projects, reducing the risk of localizers translating the wrong set of projects.
+
+# Feature explanation
+
+When creating or editing a project, a superadmin has two new options:
+
+- a simple checkbox to make the project private or public;
+- a formset controlling which users can see the project when it is private.
+
+The "approved users" formset is visible only when the project is marked as private. It has two elements: a list of users, showing who currently has access (or will upon saving), and a search input, used to look for a specific user in the database and adding that user to the list. Clicking a user in the list removes that user from the list.
+
+Administrators and managers of a project always have access to it, regarless of whether they are in the approved users list or not.
+
+On each dashboard listing projects (`/projects/`, `/{locale}/`), private projects will not appear unless the user is authenticated and is in the list of approved users.
+
+Pages related to a project (`/{project}/`, `\{locale}/{project}/`, `/{locale}/{project}/{resource}/`) will return a 404 Not Found error unless the user is authenticated and is in the list of approved users. That should also be the case for back-end endpoints used to modify data, as a user should not be able to interact at all with a private project if they are not approved for that project.
+
```

Comment:
What about Translation Memory? Strings from private projects shouldn't be visible until they get public, is it true?

Reply:
Also, how are we going to handle the Stats and their calculations for a privileged user and a non-privileged user? The stats will be different on the individual team page for a privileged and non-privileged user.

Reply:
I think it's fine to exposed strings in the Translation Memory. The goal here is not to pretend the projects do not exist, but to make sure only the right set of users can translate them.

Stats computation will be tricky indeed. @mathjazz don't we already have something similar, where stats can appear "wrong" because of hidden data?

Reply:
> I think it's fine to exposed strings in the Translation Memory. The goal here is not to pretend the projects do not exist, but to make sure only the right set of users can translate them.

That's good to know. I though private project would be secret (because "we don't want to reveal who our clients are"). We should clarify in the spec, either in the *Motivation* section or by adding the [*Out of scope*](https://docs.google.com/document/d/1W_MEYwtNn-_J0Uiu7gTSwc2B3oux75rjVtjGu-YoFi4/edit#) section.

> Stats computation will be tricky indeed. @mathjazz don't we already have something similar, where stats can appear "wrong" because of hidden data?

We exclude system projects and disabled projects from stats on most of the dashboards. Bogus stats calculations (due to which we need to fix stats after each sync job) are not related to that.